### PR TITLE
Documents undocumented parts of spring-cloud-gcp

### LINKED ISSUE #100: Add READMEs for undocumented modules
There are a few modules, like the Storage and SQL Boot starters, that don't have any documentation anywhere.

We should at least have some documentation here in the repo, either in the main README or spread around different READMEs or both.

### PR DESCRIPTION
**Pull request by joaoandremartins**
Things like property names aren't configured anywhere yet. This is one
step while we don't have a reference document with more comprehensive
documentation.

Fixes #100

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
 
 If you want to collaborate in the project, we would also love to get your Pull Requests. Before you
 start working on one, please take a look at our link:CONTRIBUTING.adoc[collaboration manual].
+
```
Comment (meltsufin) [inline README.adoc]: Three paragraphs up, the paragraph starting with "Currently, this repository contains..." needs to be updated too.

File: README.adoc [hunk 2]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+private Resource file;
+----
+
+or the Spring application context
+
+[source,java]
+----
+SpringApplication.run(...).getResource("gs://[YOUR_GCS_BUCKET]/[GCS_FILE_NAME]");
+----
+
```
Comment (meltsufin) [inline README.adoc]: You can also write to this resource:

```java
try (OutputStream os = ((WritableResource) gcsResource).getOutputStream()) {
  os.write("foo".getBytes());
}
```

File: README.adoc [hunk 3]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+== Spring Cloud GCP Storage
+
+Enables the Spring Resource abstraction to run on Google Cloud Storage.
+
+You can use the `@Value` annotation
+
+[source,java]
+----
+@Value("gs://[YOUR_GCS_BUCKET]/[GCS_FILE_NAME]")
+private Resource file;
```
Comment (meltsufin) [inline README.adoc]: I would call it `gcsResource`

File: README.adoc [hunk 4]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+
+Provides a simplified abstraction to interact with Google Cloud Pub/Sub. Using this abstraction
+allows user code to be independent from the underlying Google Cloud Pub/Sub API semantics. It's
+currently possible to publish messages to a Google Cloud Pub/Sub and to create new topics and
+subscriptions.
+
+You can refer to the `PubSubTemplate` and `PubSubAdmin` classes for more details on the
+functionality provided by this module.
+
+== Spring Cloud GCP Storage
```
Comment (meltsufin) [inline README.adoc]: I would title it as "Spring Resource Abstraction for Google Cloud Storage"

File: README.adoc [hunk 5]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+----
+
+or the Spring application context
+
+[source,java]
+----
+SpringApplication.run(...).getResource("gs://[YOUR_GCS_BUCKET]/[GCS_FILE_NAME]");
+----
+
+== Spring Integration channel adapters
```
Comment (meltsufin) [inline README.adoc]: "Pub/Sub Channel Adapters for Spring Integration" maybe?

File: README.adoc [hunk 6]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+
+Provides adapters for Spring Messaging channels to exchange messages via Google Cloud Pub/Sub.
+
+Refer to this https://spring.io/guides/gs/spring-cloud-gcp/[Spring getting started guide] to learn
+how to use the adapters.
+
+== Spring Boot GCP Core
```
Comment (meltsufin) [inline README.adoc]: "Google Cloud Core Spring Boot Starter"

File: README.adoc [hunk 7]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+http://googlecloudplatform.github.io/google-cloud-java/0.21.1/apidocs/com/google/cloud/ServiceOptions.html#getDefaultProjectId--[rules to discover the project ID].
+
+Similarly for `spring.cloud.gcp.credentials-location`, if a property is configured, the provided
+`CredentialsProvider` returns credentials for that private key. The value of
+`spring.cloud.gcp.credentials-location` is a Spring Resource, so you can specify
+https://docs.spring.io/spring/docs/current/spring-framework-reference/html/resources.html#resources-implementations[file system paths, HTTP URLs, etc.]
+If no property is configured, the provided `CredentialsProvider` returns the
+http://google.github.io/google-auth-library-java/releases/0.7.1/apidocs/com/google/auth/oauth2/GoogleCredentials.html#getApplicationDefault()[application default credentials].
+
+== Spring Boot Cloud SQL
```
Comment (meltsufin) [inline README.adoc]: "Google Cloud SQL (MySQL) Spring Boot Starter"

File: README.adoc [hunk 8]
```diff
@@ -14,3 +14,145 @@ We would love to hear from you.
...
+the ones from the provided `CredentialsProvider`, but the
+http://google.github.io/google-auth-library-java/releases/0.7.1/apidocs/com/google/auth/oauth2/GoogleCredentials.html#getApplicationDefault()[application default credentials]
+instead. However, region auto-discovery uses the credentials provided by `CredentialsProvider`.
+
+The returned `DataSource` is an implementation of
+https://brettwooldridge.github.io/HikariCP/[HikariCP]. If you want to use another kind of
+`DataSource`, feel free to use the provided `CloudSqlJdbcInfoProvider` to provide your own
+`DataSource` bean.
+
+== Spring Boot Cloud Pub/Sub
```
Comment (meltsufin) [inline README.adoc]: ...Starter

File: README.adoc (hunk 5 above)
Comment (artembilan) [inline README.adoc]: Well, usually we phrase it opposite:

> Spring Integration Channel Adapters for Google Cloud Pub/Sub

The problem here not with the chicken-eggs dilemma, but really who does what. GCP doesn't do anything for Spring Integration, really.

**Comment by joaoandremartins** [comment]
@meltsufin I also originally thought about creating different READMEs per module, but the truth is some modules wouldn't have much to say about and it could be bothersome to navigate between different READMEs. Like this, the information is resumed in a single place.

File: README.adoc (hunk 5 above)
Comment (meltsufin) [inline README.adoc]: Fair enough. :-)

**Comment by joaoandremartins** [comment]
The point about there being 3 different Pub/Sub sections resonates with me, but I think they actually are different things. I don't think we should mix them, users should be able to use the Pub/Sub channel adapters, Spring Boot starter and core module separately. Even if we're going with a coarser grain for the grouping, I think we'll still need to make a distinction between those components. Thoughts?

**Comment by meltsufin** [comment]
@joaoandremartins For each of the features, it's not completely clear what needs to be included in the `pom.xml` to get started. Maybe link to the individual modules from each of the sections?

**Comment by joaoandremartins** [comment]
I'm trying out with extracting the detail about each component to its own README. My biggest fear is that configuration information like how to configure project ID, credentials, etc. with the Core starter will be too buried in the READMEs.
Improvement ideas are welcome!

**Comment by meltsufin** [comment]
@joaoandremartins I'm kind of liking this organization of readmes. We will be putting all of this information together in the more polished single-page reference doc anyway. The readmes might just be a bit more rough.

I think one way to make the general information more prominent is to just talk about it on the top-level readme. Something like, "All starters inherit from the core which provides common configuration for project and auth. Here are the properties you configure..."

I think it's OK to have some duplication of general info across the READMEs.
Also, please pull in Jimmy to take a look and make suggestions. He might have some ideas as well.

**Comment by joaoandremartins** [comment]
@parkjam4 can you please review the contents here? Thanks!

**Comment by joaoandremartins** [comment]
We need to get this out there sooner rather than later. I'm starting to see actual user interest in this work.

I think I've been through every previous comment.

@meltsufin do you have any other comments?
@parkjam4 have you had the chance to go through this?

Thanks all

**Comment by meltsufin** [comment]
@joaoandremartins I'm fine with merging and then iterating on it post-merge.

**Comment by joaoandremartins** [comment]
Jimmy is reviewing the docs as we speak. I'll let him finish and make any fixes. I could use your approval afterwards.

File: spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/README.adoc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+[source]
+----
+dependencies {
+    compile group: 'org.springframework.cloud', name: 'spring-cloud-gcp-starter-sql', version: '1.0.0.BUILD-SNAPSHOT'
+}
+----
+
+
+This starter provides a couple of configuration sources: `CloudSqlJdbcInfoProvider` and
```
Comment (parkjam4) [inline spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/README.adoc]: replace "a couple of" with "the"

File: spring-integration-gcp/README.adoc [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    compile group: 'org.springframework.cloud', name: 'spring-integration-gcp', version: '1.0.0.BUILD-SNAPSHOT'
+}
+----
+
+
+Provides adapters for Spring Messaging channels to exchange messages via Google Cloud Pub/Sub. Makes
+extensive usage of the constructs provided by the Spring Cloud GCP Pub/Sub module, like
+`PubSubTemplate`.
+
+Refer to this https://spring.io/guides/gs/spring-cloud-gcp/[Spring getting started guide] to learn
```
Comment (parkjam4) [inline spring-integration-gcp/README.adoc]: replace "this" with "the"

File: spring-integration-gcp/README.adoc [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+[source]
+----
+dependencies {
+    compile group: 'org.springframework.cloud', name: 'spring-integration-gcp', version: '1.0.0.BUILD-SNAPSHOT'
+}
+----
+
+
+Provides adapters for Spring Messaging channels to exchange messages via Google Cloud Pub/Sub. Makes
+extensive usage of the constructs provided by the Spring Cloud GCP Pub/Sub module, like
```
Comment (parkjam4) [inline spring-integration-gcp/README.adoc]: replace "usage" with "use"

File: README.adoc [hunk 12]
```diff
@@ -4,13 +4,60 @@ This project aims at making it as seamless as possible for Spring users to run t
 Google Cloud Platform. The project website is http://cloud.spring.io/spring-cloud-gcp/ (under
 development).
 
-Currently, this repository contains Spring Integration inbound and outbound channel adapters which
-use Google Cloud Pub/Sub, as well as a Google Cloud Pub/Sub Spring abstraction, i.e.,
-`PubSubTemplate`.
+Currently, this repository provides support for:
```
Comment (parkjam4) [inline README.adoc]: Suggestion:
in Line 3, change "This project aims at making it as seamless as possible for Spring users..."
to
"This project makes it easier for Spring users..."

**Comment by joaoandremartins** [comment]
I applied fixes for Jimmy's suggestion. @meltsufin can you please approve if you're happy?

**Pull request merged**